scene/gui/text_edit_menu.cpp

```diff
@@ -0,0 +1,214 @@
+	idx = m_menu->get_item_index(m_id);                                                                                  \
+	if (idx >= 0) {                                                                                                      \
+		m_menu->set_item_accelerator(idx, p_shortcut_keys_enabled ? _get_menu_action_accelerator(m_action) : Key::NONE); \
+	}
+
+#define MENU_ITEM_DISABLED(m_menu, m_id, m_disabled) \
+	idx = m_menu->get_item_index(m_id);              \
+	if (idx >= 0) {                                  \
+		m_menu->set_item_disabled(idx, m_disabled);  \
+	}
+
+#define MENU_ITEM_CHECKED(m_menu, m_id, m_checked) \
+	idx = m_menu->get_item_index(m_id);            \
+	if (idx >= 0) {                                \
+		m_menu->set_item_checked(idx, m_checked);  \
+	}
+
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_CUT, "ui_cut", !p_editable)
+	MENU_ITEM_ACTION(menu, MENU_COPY, "ui_copy")
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_PASTE, "ui_paste", !p_editable)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_SELECT_ALL, "ui_text_select_all", !p_selecting_enabled)
+	MENU_ITEM_DISABLED(menu, MENU_CLEAR, !p_editable)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_UNDO, "ui_undo", !p_editable || !p_has_undo)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_REDO, "ui_redo", !p_editable || !p_has_redo)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_INHERITED, p_text_direction == Control::TEXT_DIRECTION_INHERITED)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_AUTO, p_text_direction == Control::TEXT_DIRECTION_AUTO)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_LTR, p_text_direction == Control::TEXT_DIRECTION_LTR)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_RTL, p_text_direction == Control::TEXT_DIRECTION_RTL)
+	MENU_ITEM_CHECKED(menu, MENU_DISPLAY_UCC, p_draw_control_chars)
+	MENU_ITEM_DISABLED(menu, MENU_SUBMENU_INSERT_UCC, !p_editable)
+
+#undef MENU_ITEM_ACTION_DISABLED
+#undef MENU_ITEM_ACTION
+#undef MENU_ITEM_DISABLED
+#undef MENU_ITEM_CHECKED
+}
+
+void TextEditMenu::popup_at(const Point2 &p_point, bool p_grab_focus) {
+	if (menu) {
+		menu->set_position(p_point);
+		menu->reset_size();
+		menu->popup();
+		if (p_grab_focus) {
+			menu->grab_focus();
+		}
+	}
+}
+
+bool TextEditMenu::has_focus() {
+	return menu && menu->has_focus();
+}
+
+void TextEditMenu::set_text_direction(Control::TextDirection p_text_direction) {
+	if (menu_dir) {
+		menu_dir->set_item_checked(menu_dir->get_item_index(MENU_DIR_INHERITED), p_text_direction == Control::TEXT_DIRECTION_INHERITED);
+		menu_dir->set_item_checked(menu_dir->get_item_index(MENU_DIR_AUTO), p_text_direction == Control::TEXT_DIRECTION_AUTO);
+		menu_dir->set_item_checked(menu_dir->get_item_index(MENU_DIR_LTR), p_text_direction == Control::TEXT_DIRECTION_LTR);
+		menu_dir->set_item_checked(menu_dir->get_item_index(MENU_DIR_RTL), p_text_direction == Control::TEXT_DIRECTION_RTL);
+	}
+}
+
+void TextEditMenu::set_draw_control_chars(bool p_draw_control_chars) {
+	if (menu && menu->get_item_index(MENU_DISPLAY_UCC) >= 0) {
```

Comment:
This was here originally, but I wonder what the second check is for. That option exists unconditionally.

Reply:
Not sure about 2nd check, I assume just being extra cautious (replaced 1st check with ERR_FAIL_ like you suggested). I left it in there for safety.

Reply:
Unless there's something you want me to do, I'll leave it for you to resolve. My best guess is the code is there in case someone removes that entry from the Popup menu (the menu is open for people to add entries to it, like filter option in the Filter LineEdit, so I assume people would theoretically be able to remove entries from it too).